ganttproject/src/main/java/net/sourceforge/ganttproject/storage/SqlProjectDatabaseImpl.kt

```diff
@@ -204,51 +224,83 @@ class SqlProjectDatabaseImpl(private val dataSource: DataSource) : ProjectDataba
 
   /** Add update query and save its xlog in the current transaction. */
   @Throws(ProjectDatabaseException::class)
-  internal fun update(h2Query: String, postgresQuery: String) = withLog({ "Failed to execute update" }, h2Query, postgresQuery)
+  internal fun update(queries: List<SqlQuery>, undoQueries: List<SqlUndoQuery>) = withLog(queries, undoQueries)
 }
-internal data class Query(
-  val errorMessage: () -> String,
+
+data class SqlQuery(
   val sqlStatementH2: String,
   val sqlStatementPostgres: String
 )
 
+typealias SqlUndoQuery = SqlQuery
+
 class TransactionImpl(private val database: SqlProjectDatabaseImpl, private val title: String): ProjectDatabaseTxn {
-  internal val statements = mutableListOf<Query>()
+  private val statements = mutableListOf<SqlQuery>()
+  private val undoStatements = mutableListOf<SqlQuery>()
+
+  private var isCommitted: Boolean = false
 
   override fun commit() {
-    database.commitTransaction(this)
+    if (isCommitted) throw ProjectDatabaseException("Transaction is already committed")
+    database.commitTransaction(statements)
+    isCommitted = true
+  }
+
+  override fun undo() {
+    if (!isCommitted) throw ProjectDatabaseException("Cannot undo uncommitted transaction")
+    database.commitTransaction(undoStatements)
+  }
+
+  override fun redo() {
+    if (!isCommitted) throw ProjectDatabaseException("Cannot redo uncommitted transaction")
+    database.commitTransaction(statements)
   }
 
-  internal fun add(query: Query) {
+  internal fun add(query: SqlQuery, undoQuery: SqlQuery) {
+    if (isCommitted) throw ProjectDatabaseException("Txn was already committed")
     statements.add(query)
+    undoStatements.add(0, undoQuery)
+  }
+
+  internal fun add(queries: List<SqlQuery>, undoQueries: List<SqlUndoQuery>) {
+    if (isCommitted) throw ProjectDatabaseException("Txn was already committed")
+    statements.addAll(queries)
+    undoStatements.addAll(0, undoQueries)
```

Comment:
Hmm, for now, the only reason why `add` accepts lists should be the fact that setCustomProperties can generate different number of statements for commit & undo, AND we want to preserve delete & merge order in undo. The case should be the only place which passes a list with the size >= 1. Does it sound better to build custom properties undo list, adding merge first and delete later intentionally, so that we could reverse? :)
Maybe we should have a `preserveOrder` flag, which is `false` by default, to make the case clear.

Reply:
I am sure that there will be many cases with multiple statements. E.g. if you change the project start date in a sample project (House Building Sample) e.g. by moving the start milestone task, you will have a lot of task date changes. They not necessarily need the order, but it is better not to guess if they need the order or not.